### ISSUE TITLE
ActiveStorage::MirrorService uploads to mirrors inline instead of using an async ActiveStorage::MirrorJob

### ISSUE THREAD

**Issue by feliperaul** [rails/rails#47918]
`ActiveStorage::Mirror`  is really nice to have. However, on Rails 7.0.4.3 (and `master` as well), I realized that when you define a `Mirror` service on `storage.yml` , the upload to the mirror actually happens inline, instead of using `ActiveStorage::MirrorJob`.

This actually slows down any uploads to the service, because it is performed n times (n being the number of mirrors), inline, instead of using the `ActiveStorage::MirrorJob`. 

I don't think this is by design, since the `ActiveStorage::MirrorJob` actually gets enqueued right away, but it's actually useless: when it runs, there are no more uploads left to do in any mirror service, and that's the only moment the job is enqueued. 

What actually happens is:

1. When a Blob is created,`ActiveStorage::Blob#upload` will be called, which calls the service's `#upload`, in this case, `ActiveStorage::MirrorService#upload`. 

2. The `ActiveStorage::MirrorService` `#upload` will do the upload inline for each service:

https://github.com/rails/rails/blob/be287ac0d5000e667510faba0dbf28ffdb7c5e35/activestorage/lib/active_storage/service/mirror_service.rb#L37-L42

That's why, on the Rails log, you see two uploads one right after the other:

```
S3 Storage (1159.8ms) Uploaded file to key: nqg705we2u70olmnplgfujbwpb63 (checksum: GE6tZIm49SutC3TY9kV4Uw==)
S3 Storage (898.7ms) Uploaded file to key: nqg705we2u70olmnplgfujbwpb63 (checksum: GE6tZIm49SutC3TY9kV4Uw==)
[ActiveJob] Enqueued ActiveStorage::MirrorJob (Job ID: 3eaeaa5e-6156-4a37-b048-1e9a96484867) to Sidekiq(low) with arguments: "nqg705we2u70olmnplgfujbwpb63", {:checksum=>"GE6tZIm49SutC3TY9kV4Uw=="}
```

As you can see, an `ActiveStorage::MirrorJob` was also enqueued, due to the `after_create_commit :mirror_blob_later` on `ActiveStorage::Attachment`. However, it's enqueued only after the 2 uploads were already performed.

And when that Job actually runs, it will simply invoke `#mirror` on the `ActiveStorage::MirrorService`, which has this code:

https://github.com/rails/rails/blob/be287ac0d5000e667510faba0dbf28ffdb7c5e35/activestorage/lib/active_storage/service/mirror_service.rb#L55-L67

By that moment in time, `mirrors_in_need_of_mirroring` will always be blank (I confirmed), unless in the edge-case that between the upload happening and the job actually running a new mirror was added to the service configuration, which is probably not the desired use case here.

In a nutshell, as it is:

- The mirror configuration slows down uploads, since it's performing all of them inline; if there are multiple mirrors configured, this compounds.
- An `ActiveStorage::MirrorJob` gets enqueued, but it's pretty much useless; and it's not even free, because it will aways fire S3 requests because it will check for the key presence in each one of the mirrors (which they will already be). 

The Mirror documentation already makes no promises of ensuring mirroring happens sucessfully:

> Mirroring is not atomic. It is possible for an upload to succeed on the primary service and fail on any of the subordinate services. Before going all-in on a new service, verify that all files have been copied.

If that's the case, shouldn't we just upload once to the primary, and let `ActiveStorege::MirrorService` do it's job and mirror asynchronously to the other storages later on?

Maybe only the `local` service could have its mirroring done inline. 

Mentioning @georgeclaghorn because he authored the original code and maybe I got this all wrong.

**Comment by ghiculescu** [comment]
This sounds like a bug to me too. Would you be interested in making a PR to fix?

**Comment by feliperaul** [comment]
Great, I'll give it a shot.

Since I'm at it, I'd like to improve how to sync new mirrors. I think that everyone that will use the mirror feature will have this need of backfilling the new mirror with all the blobs already in existance. 

There's a [stackoverflow question](https://stackoverflow.com/questions/52684443/how-to-sync-new-activestorage-mirrors) and the most upvoted question is wrong; `blob.mirror_later` will do nothing for existing blobs that were created with a service that was not a Mirror service.
